File: samples/c/ftl/tibmapset.c
/*
 * Copyright (c) 2010-2017 TIBCO Software Inc.
 * Licensed under a BSD-style license. Refer to [LICENSE]
 * For more information, please contact:
 * TIBCO Software Inc., Palo Alto, California, USA
 */

#include <stdlib.h>
#include <stdio.h>
#include <stddef.h>
#include <string.h>
#include <ctype.h>

#include "tib/ftl.h"

#include "tibaux.h"

#define DEFAULT_URL "http://localhost:8080"
#define DEFAULT_SECURE_URL "https://localhost:8080"

const char      *appName        = "tibmapset";
const char      *realmServer    = NULL;
const char      *mapEndpoint    = "tibmapset-endpoint";
const char      *trcLevel       = NULL;
const char      *user           = "guest";
const char      *password       = "guest-pw";
const char      *identifier     = NULL;
const char      *formatName     = "Format-1";
const char      *secondary      = NULL;
const char      *mapName        = "tibmap";
char            *keys           = "";
char            *keysStart      = NULL;
const char      *keyList        = "key1,key2,key3";
tibbool_t       removeMap       = tibfalse;
tibbool_t       trustAll        = tibfalse;
const char      *trustFile      = NULL;
const char      *clientLabel    = NULL;
tibMessage      msg             = NULL;
tibRealm        realm;
tibint32_t      count = 0;

void
usage(
    int         argc,
    char        **argv)
{
    printf("Usage: %s [options] url\n", argv[0]);
    printf("Default url is "DEFAULT_URL"\n");
    printf("    -a,  --application <name>\n"
           "         --client-label <label>\n"
           "    -e,  --endpoint <name>\n"
           "    -f,  --format <name>\n"
           "    -h,  --help\n"
           "    -id, --identifier <name>\n"
           "    -l,  --keyList <string>[,<string>]\n"
           "    -n,  --mapname <name>\n"
           "    -p,  --password <string>\n"
           "    -u,  --user <string>\n"
           "    -s,  --secondary <string>\n"
           "    -t,  --trace <level>\n"
           "            where <level> can be:\n"
           "                  off, severe, warn, info, debug, verbose\n"
           "         --trustall\n"
           "         --trustfile <file>\n");

    exit(1);
}

void
parseArgs(int argc, char** argv)
{
    int    i = 1;

    tibAux_InitGetArgs(argc, argv, usage);

    for (i = 1;  i < argc;  i++)
    {
        if (tibAux_GetString(&i, "--application", "-a", &appName))
            continue;
        if (tibAux_GetString(&i, "--client-label", "--client-label", &clientLabel))
            continue;
        if (tibAux_GetString(&i, "--endpoint", "-e", &mapEndpoint))
            continue;
        if (tibAux_GetString(&i ,"--format", "-f", &formatName))
            continue;
        if (tibAux_GetFlag(&i, "--help", "-h"))
            usage(argc, argv);
        if (tibAux_GetString(&i, "--identifier", "-id", &identifier))
            continue;
        if (tibAux_GetString(&i, "--keyList", "-l", &keyList))
            continue;
        if (tibAux_GetString(&i, "--mapname", "-n", &mapName))
            continue;
        if (tibAux_GetString(&i, "--password", "-p", &password))
            continue;
        if (tibAux_GetString(&i, "--secondary", "-s", &secondary))
            continue;
        if (tibAux_GetString(&i, "--trace", "-t", &trcLevel))
            continue;
        if (tibAux_GetString(&i, "--user", "-u", &user))
            continue;
        if (tibAux_GetFlag(&i, "--trustall", "--trustall")) {
            if (trustFile)
            {
                printf("cannot specify both --trustall and --trustfile\n");
                usage(argc, argv);
            }
            trustAll = tibtrue;
            continue;
        }
        if (tibAux_GetString(&i, "--trustfile", "--trustfile", &trustFile))
        {
            if (trustAll)
            {
                printf("cannot specify both --trustall and --trustfile\n");
                usage(argc, argv);
            }
            continue;
        }
        if (strncmp(argv[i], "-", 1)==0)
        {
            printf("invalid option: %s\n", argv[i]);
            usage(argc, argv);
        }
        realmServer = argv[i];
    }

    printf("Invoked as: ");
    for (i = 0;  i < argc;  i++)
        printf("%s ", argv[i]);
    printf("\n");

    if (keyList)
    {
        keysStart = tibAux_strdup(keyList);
        keys  = keysStart;
    }
}

void
memFail(
    const char          *obj)
{
    printf("memory allocation failed for %s\n", obj);
    exit(-1);
}

void
initMessage(
    tibEx               ex,
    char                *key)
{
    if (msg == NULL)
        msg = tibMessage_Create(ex, realm, formatName);
    else
        tibMessage_ClearAllFields(ex, msg);

    tibMessage_SetLong(ex, msg, "My-Long", ++count);
}

int
main(int argc, char** argv)
{
    tibEx               ex;
    tibProperties       props;
    tibMap              tibmap;

    printf("#\n# %s\n#\n# %s\n#\n", argv[0], tib_Version());

    parseArgs(argc, argv);

    ex = tibEx_Create();
    if (ex == NULL)
        memFail("ex");

    tib_Open(ex, TIB_COMPATIBILITY_VERSION);
    CHECK(ex);

    // Set global trace to specified level
    if (trcLevel != NULL)
        tib_SetLogLevel(ex, trcLevel);

    // Get a single realm per process
    props = tibProperties_Create(ex);
    tibProperties_SetString(ex, props,
                            TIB_REALM_PROPERTY_STRING_USERNAME, user);
    tibProperties_SetString(ex, props,
                            TIB_REALM_PROPERTY_STRING_USERPASSWORD, password);
    if (identifier)
        tibProperties_SetString(ex, props,
                                TIB_REALM_PROPERTY_STRING_APPINSTANCE_IDENTIFIER, identifier);
    if (secondary)
        tibProperties_SetString(ex, props,
                                TIB_REALM_PROPERTY_STRING_SECONDARY_SERVER, secondary);

    if (trustAll || trustFile)
    {
        if (!realmServer)
            realmServer = DEFAULT_SECURE_URL;

        if (trustAll)
        {
            tibProperties_SetLong(ex, props, TIB_REALM_PROPERTY_LONG_TRUST_TYPE,
                                        TIB_REALM_HTTPS_CONNECTION_TRUST_EVERYONE);
        }
        else
        {
            tibProperties_SetLong(ex, props, TIB_REALM_PROPERTY_LONG_TRUST_TYPE,
                    TIB_REALM_HTTPS_CONNECTION_USE_SPECIFIED_TRUST_FILE);
            tibProperties_SetString(ex, props, TIB_REALM_PROPERTY_STRING_TRUST_FILE, trustFile);
        }
    }
    else
    {
        if (!realmServer)
            realmServer = DEFAULT_URL;
    }

    // set the client label
    if (clientLabel != NULL)
        tibProperties_SetString(ex, props, TIB_REALM_PROPERTY_STRING_CLIENT_LABEL, clientLabel);
    else
        tibProperties_SetString(ex, props, TIB_REALM_PROPERTY_STRING_CLIENT_LABEL, argv[0]);

    realm = tibRealm_Connect(ex, realmServer, appName, props);
    CHECK(ex);
    tibProperties_Destroy(ex, props);

    // create a dynamic map in tibstore that will store the key-value 
    // pairs
    tibmap = tibRealm_CreateMap(ex, realm, mapEndpoint, mapName, NULL);
    CHECK(ex);

    if (keyList)
    {
        char* key = tibAux_strtok_r(keys, ",", &keys);
    
        do
        {
            printf("setting value for key %s\n", key); fflush(stdout);
            initMessage(ex, key);

            tibMap_Set(ex, tibmap, key, msg);
            CHECK(ex);
        }
        while ((key = tibAux_strtok_r(keys, ",", &keys)));
    }

    // close the map
    tibMap_Close(ex, tibmap);

    tibRealm_Close(ex, realm);
    tib_Close(ex);

    // Check if it worked before destroying the exception
    CHECK(ex);
    tibEx_Destroy(ex);

    return 0;
}

File: samples/java/ftl/TibMapSet.java
/*
 * Copyright (c) 2010-2017 TIBCO Software Inc.
 * Licensed under a BSD-style license. Refer to [LICENSE]
 * For more information, please contact:
 * TIBCO Software Inc., Palo Alto, California, USA
 */

/*
 * This is a sample of a basic java program which demonstrates the TibMap API that
 * sets the requested number of keys with corresponding values in the 'tibmap' at the persistence server.
 */

package com.tibco.ftl.samples;

import java.util.*;
import com.tibco.ftl.*;
import com.tibco.ftl.exceptions.*;

public class TibMapSet implements NotificationHandler
{
    String          appName     = "tibmapset";
    String          realmServer = "http://localhost:8080";
    String          mapEndpoint = "tibmapset-endpoint";
    String          formatName  = "Format-1";
    String          trcLevel    = null;
    String          user        = "guest";
    String          password    = "guest-pw";
    String          identifier  = null;
    String          secondary   = null;
    String          mapName     = "tibmap";
    static final String DEFAULT_URL = "http://localhost:8080";
    static final String DEFAULT_SECURE_URL = "https://localhost:8080";
    String          keyList     = "key1,key2,key3";
    String          clientLabel = null;

    Realm           realm       = null;
    TibMap          map         = null;
    Message         msg         = null;
    int             count       = 0;
    boolean         trustAll    = false;
    String          trustFile   = null;
    

    String          usageString =
        "TibMapSet [options] url\n" +
        "Default url is " + realmServer + "\n" +
        "  where options can be:\n" +
        "    -a, --application <name>\n" +
        "    -c, --count <int>\n" +
        "    -cl, --client-label <string>\n" +
        "    -e, --endpoint <name>\n" +
        "    -f, --format <name>\n" +
        "    -h, --help\n" +
        "    -id, --identifier <string>\n" +
        "    -l,  --keyList <string>[,<string>]\n" +
        "    -n,  --mapname <name>\n" +
        "    -p, --password <string>\n" +
        "    -s, --secondary <string>\n" +
        "    -t, --trace <level>\n" +
        "           where <level> can be:\n" +
        "                 off, severe, warn, info, debug, verbose\n" +
        "        --trustall\n" +
        "        --trustfile <file>\n" +
        "    -u, --user <string>\n";

    public TibMapSet(String[] args)
    {
        System.out.printf("#\n# %s\n#\n# %s\n#\n",
                          this.getClass().getName(),
                          FTL.getVersionInformation());
        
        parseArgs(args);
    }

    public void usage()
    {
        System.out.println(usageString);
        System.exit(0);
    }

    public void parseArgs(String args[])
    {
        int i           = 0;
        int argc        = args.length;
        String s        = null;
        int n           = 0;
        TibAux aux      = new TibAux(args, usageString);

        for (i = 0; i < argc; i++)
        {
            s = aux.getString(i, "--application", "-a");
            if (s != null) {
                appName = s;
                i++;
                continue;
            }
            s = aux.getString(i, "--client-label", "-cl");
            if (s != null) {
                clientLabel = s;
                i++;
                continue;
            }
            n = aux.getInt(i, "--count", "-c");
            if (n > -1) {
                count = n;
                i++;
                continue;
            }
            s = aux.getString(i, "--endpoint", "-e");
            if (s != null) {
                mapEndpoint = s;
                i++;
                continue;
            }
            s = aux.getString(i, "--format", "-f");
            if (s != null)
            {
                formatName = s;
                i++;
                continue;
            }
            if (aux.getFlag(i, "--help", "-h")) {
                usage();
            }
            s = aux.getString(i, "--identifier", "-id");
            if (s != null) {
                identifier = s;
                i++;
                continue;
            }
            s = aux.getString(i, "--keyList", "-l");
            if (s != null) {
                keyList = s;
                i++;
                continue;
            }
            s = aux.getString(i, "--mapname", "-n");
            if (s != null) {
                mapName = s;
                i++;
                continue;
            }
            s = aux.getString(i, "--password", "-p");
            if (s != null) {
                password = s;
                i++;
                continue;
            }
            s = aux.getString(i, "--secondary", "-s");
            if (s != null) {
                secondary = s;
                i++;
                continue;
            }
            s = aux.getString(i, "--trace", "-t");
            if (s != null) {
                trcLevel = s;
                i++;
                continue;
            }
            if (aux.getFlag(i, "--trustall", "--trustall")) {
                if (trustFile != null)
                {
                    System.out.println("cannot specify both --trustall and --trustfile");
                    usage();
                }
                trustAll = true;
                continue;
            }
            s = aux.getString(i, "--trustfile", "--trustfile");
            if (s != null)
            {
                trustFile = s;
                if (trustAll)
                {
                    System.out.println("cannot specify both --trustall and --trustfile");
                    usage();
                }
                i++;
                continue;
            }
            s = aux.getString(i, "--user", "-u");
            if (s != null) {
                user = s;
                i++;
                continue;
            }
            if (args[i].startsWith("-", 0)) {
                System.out.println("invalid option: "+args[i]);
                usage();
            }
            realmServer = args[i];
        }

        System.out.print("Invoked as: TibMapSet ");
        for (i = 0; i < argc; i++)
            System.out.print(args[i]+" ");
        System.out.println("");
    }
    
    public void onNotification(int type, String reason)
    {
        if (type == NotificationHandler.CLIENT_DISABLED)
        {
            System.out.println("application administratively disabled: " + reason);
            System.out.println("exiting");
            System.exit(1);
        }
        else
        {
            System.out.println("notification type " + type + ": " + reason);
        }
    }

    public void initMessage(String key) throws FTLException
    {
        // Create sender object and message to be sent
        if (msg == null)
            msg = realm.createMessage(formatName);
        else
            msg.clearAllFields();

        msg.setLong("My-Long", ++count);
    }

    public int work() throws FTLException
    {
        TibProperties   props;
        
        // Set global trace to specified level
        if (trcLevel != null)
            FTL.setLogLevel(trcLevel);

        // Get a single realm per process
        props = FTL.createProperties();
        props.set(Realm.PROPERTY_STRING_USERNAME, user);
        props.set(Realm.PROPERTY_STRING_USERPASSWORD, password);
        if (identifier != null)
            props.set(Realm.PROPERTY_STRING_APPINSTANCE_IDENTIFIER, identifier);
        if (secondary != null)
            props.set(Realm.PROPERTY_STRING_SECONDARY_SERVER, secondary);

        if (trustAll == true || trustFile != null)
        {
            if (realmServer == null)
                realmServer = TibMapSet.DEFAULT_SECURE_URL;

            if (trustAll)
                props.set(Realm.PROPERTY_LONG_TRUST_TYPE, Realm.HTTPS_CONNECTION_TRUST_EVERYONE);
            else
            {
                props.set(Realm.PROPERTY_LONG_TRUST_TYPE, Realm.HTTPS_CONNECTION_USE_SPECIFIED_TRUST_FILE);
                props.set(Realm.PROPERTY_STRING_TRUST_FILE, trustFile);
            }
        }
        else
        {
            if (realmServer == null)
                realmServer = TibMapSet.DEFAULT_URL;
        }

        if (clientLabel != null)
            props.set(Realm.PROPERTY_STRING_CLIENT_LABEL, clientLabel);
        else
            props.set(Realm.PROPERTY_STRING_CLIENT_LABEL, this.getClass().getSimpleName());

        realm = FTL.connectToRealmServer(realmServer, appName, props);
        props.destroy();

        // Set this object as the NotificationHandler to receive notifications
        realm.setNotificationHandler(this);
        map = realm.createMap(mapEndpoint, mapName, null);

        StringTokenizer tok = new StringTokenizer(keyList);
        try
        {
            while(tok.hasMoreElements())
            {
                String key = (String)tok.nextToken(",");
                System.out.println("setting value for key " + key);
                initMessage(key);

                map.set(key, msg);
            }
        }
        catch (FTLClientShutdownException ftle)
        {
            // Something unusual happened and the Realm was shutdown. Give the
            // notification handler a chance to run so that we can log if this
            // is an administrative action.
            try
            {
                Thread.sleep(1000);
            }
            catch (InterruptedException ie)
            {
                // ignore
            }

            ftle.printStackTrace();
        }

        // Clean up
        map.close();
        realm.close();

        return 0;
    }
    
    public static void main(String[] args)
    {
        TibMapSet s  = new TibMapSet(args);
        try
        {
            s.work();
        }
        catch (Throwable e)
        {
            e.printStackTrace();
        }
    }
}
